src/EasyAbp.BookingService.Domain/EasyAbp/BookingService/AssetOccupancyProviders/AssetOccupancyProviderBase.cs

```diff
@@ -64,6 +64,62 @@ protected AssetOccupancyProviderBase(IServiceProvider serviceProvider)
     {
         var periodScheme = await GetEffectivePeriodSchemeAsync(targetDate, asset, categoryOfAsset);
 
+        return await InternalGetPeriodsAsync(asset, categoryOfAsset, periodScheme, targetDate, currentDateTime);
+    }
+
+    [UnitOfWork]
+    public virtual async Task<List<PeriodOccupancyModel>> GetPeriodsAsync(AssetCategory category,
+        Guid? periodSchemeId,
+        DateTime targetDate,
+        DateTime? currentDateTime = default)
+    {
+        var assets =
+            await AssetRepository.GetListAsync(x => x.AssetCategoryId == category.Id && !x.Disabled);
+
+        var effectivePeriodScheme = await GetEffectivePeriodSchemeAsync(periodSchemeId, category);
+
+        var models = effectivePeriodScheme.Periods
+            .Select(x =>
+                new PeriodOccupancyModel(targetDate,
+                    x.StartingTime,
+                    x.GetEndingTime(),
+                    effectivePeriodScheme.Id,
+                    x.Id,
+                    0, 0))
+            .ToList();
+
+        foreach (var asset in assets)
+        {
+            var assetPeriodScheme = await GetEffectivePeriodSchemeAsync(targetDate, asset, category);
+            if (effectivePeriodScheme.Id != assetPeriodScheme.Id)
+            {
+                continue;
+            }
+
+            var assetPeriodOccupancyModels =
+                await InternalGetPeriodsAsync(asset, category, assetPeriodScheme, targetDate, currentDateTime);
+
+            var assetPeriodOccupancyModelDictionary = assetPeriodOccupancyModels.ToDictionary(x => x.PeriodId);
+            foreach (var periodOccupancyModel in models)
+            {
+                var assetPeriodOccupancyModel = assetPeriodOccupancyModelDictionary[periodOccupancyModel.PeriodId];
+                if (periodOccupancyModel.AvailableVolume < assetPeriodOccupancyModel.AvailableVolume)
+                {
+                    periodOccupancyModel.TotalVolume = assetPeriodOccupancyModel.TotalVolume;
+                    periodOccupancyModel.AvailableVolume = assetPeriodOccupancyModel.AvailableVolume;
+                }
```

Comment:
Maybe we should calculate max volume of every period instead of getting first available volume of every period? @gdlcf88

Reply:
And yes, it may come with some performance issue, so do we need asset occupancy cache in category search? @gdlcf88

Reply:
> Maybe we should calculate max volume of every period instead of getting first available volume of every period?

`IAssetInCategorySelector` and `DefaultAssetInCategorySelector` could be a good solution. For the default implementation we only care about the `Priority`.

Reply:
> And yes, it may come with some performance issue, so do we need asset occupancy cache in category search? @gdlcf88

I think we shouldn't cache an entity.